src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,267 @@
+        // link the alloc id to the actual allocation
+        let alloc = tcx.intern_const_alloc(alloc);
+        self.leftover_relocations.extend(alloc.relocations.iter().map(|&(_, ((), reloc))| reloc));
+        tcx.alloc_map.lock().set_alloc_id_memory(ptr.alloc_id, alloc);
+        Ok(None)
+    }
+}
+
+impl<'rt, 'a, 'mir, 'tcx>
+    ValueVisitor<'a, 'mir, 'tcx, CompileTimeInterpreter<'a, 'mir, 'tcx>>
+for
+    InternVisitor<'rt, 'a, 'mir, 'tcx>
+{
+    type V = MPlaceTy<'tcx>;
+
+    #[inline(always)]
+    fn ecx(&self) -> &CompileTimeEvalContext<'a, 'mir, 'tcx> {
+        &self.ecx
+    }
+
+
+    fn visit_aggregate(
+        &mut self,
+        mplace: MPlaceTy<'tcx>,
+        _fields: impl Iterator<Item=EvalResult<'tcx, Self::V>>,
+    ) -> EvalResult<'tcx> {
+        if let Some(def) = mplace.layout.ty.ty_adt_def() {
+            if Some(def.did) == self.ecx.tcx.lang_items().unsafe_cell_type() {
+                // we are crossing over an `UnsafeCell`, const qualif ensures that this can only
+                // exist in the base constant, not behind references. We may relax this in
+                // the future
+                match self.mode {
+                    // static mut is unsound anyway
+                    InternMode::StaticMut => {},
+                    // if we are in a frozen static (so behind a refernce), const qualif must
+                    // have a bug.
+                    InternMode::Const { frozen } |
+                    InternMode::Static { frozen } => if frozen {
+                        bug!("UnsafeCell in constant behind reference");
+                    },
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn visit_primitive(&mut self, mplace: MPlaceTy<'tcx>) -> EvalResult<'tcx> {
+        // Handle Reference types, as these are the only relocations supported by const eval.
+        // Raw pointers (and boxes) are handled by the `leftover_relocations` logic.
+        let ty = mplace.layout.ty;
+        if let ty::Ref(..) = ty.sty {
+            let value = self.ecx.read_immediate(mplace.into())?;
+            // Handle trait objects
+            if let Ok(meta) = value.to_meta() {
+                let layout = self.ecx.layout_of(ty.builtin_deref(true).unwrap().ty)?;
+                if layout.is_unsized() {
+                    if let ty::Dynamic(..) = self.ecx.tcx.struct_tail(layout.ty).sty {
+                        if let Ok(vtable) = meta.unwrap().to_ptr() {
+                            match self.mode {
+                                | InternMode::Static { frozen: true }
+                                | InternMode::Const { frozen: true }
+                                => {},
+                                | InternMode::StaticMut
+                                | InternMode::Static { frozen: false }
+                                | InternMode::Const { frozen: false }
+                                => bug!("vtables are behind references and should thus be frozen"),
```

Comment:
Seems like this and the other frozen check could be simplified by adding a `InternMode::frozen` method.

Reply:
I completely refactored this, so it's just a comparison against `Mutability::Immutable` nowadays